### PR TITLE
xcode-wrapper: allow xcodebuild to fail in version check

### PR DESCRIPTION
**Pull request by mroi**
With the current beta of Xcode 16.0, the `xcode-wrapper` derivation fails with error code 141. No additional output is generated. I traced this back to `xcodebuild -version` failing when its output is piped to `head`. I think this is because head closes xcodebuild’s stdout after the first line, which xcodebuild does not like.

## Description of changes

Fixed by allowing xcodebuild to fail within the pipe used for version number extraction.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by jakubgs — CHANGES_REQUESTED** [review]
No, why would you do that?

**Comment by mroi** [comment]
With the Xcode 16 beta, the Nixpkgs `xcode-wrapper` fails to build.

`xcodebuild` does indeed return a nonzero exit code when output is piped to `head -n1`. I guess it does not handle the fact that its stdout is closed when it wants to output its second line. We could try and get Apple to fix this, but I thought this can be an acceptable fix as we don’t really care about the `xcodebuild` exit code as long as the version number it outputs is the expected one.

**Comment by jakubgs** [comment]
As always apple coming out with the most mind boggling behaviors that make no sense.

I'd rather not just ignore Xcode errors. Could we not just pipe it through `cat` first so it reads all lines?
```sh
    currVer=$($out/bin/xcodebuild -version | cat | head -n1)
```
Based on your description that should not fail right?

**Comment by mroi** [comment]
I investigated some more and the problem is not specific to `xcodebuild` or macOS at all. [It also happens on Linux with other tools.](https://stackoverflow.com/questions/19120263/)

tl;dr: When piping output to a tool that may exit prematurely, without consuming all its input, the process on the left of the pipe receives a `SIGPIPE` and the standard reaction leads to exit code 141.

So piping to `head -n1` within a script with pipefail enabled is brittle and may break as it does here. This should work reliably:

```
versionInfo=$($out/bin/xcodebuild -version)
currVer="${versionInfo%%$'\n'*}"  # Xcode version is in the first line
```

Do you think this is better than masking the exit code with `|| true`? It should be more robust, as it checks the exit code, but conveys the intention a little less clearly.

**Comment by mroi** [comment]
This also works, maybe the best compromise? Do you have a preference?

```
currVer=$(head -n1 <(xcodebuild -version))
```

**Comment by jakubgs** [comment]
But I don't see `pipefail` enabled in that `buildCommand`. Where are you getting `pipefail` from there?

Also, I see no such behavior from Xcode 15.1.

**Comment by mroi** [comment]
> But I don't see `pipefail` enabled in that `buildCommand`. Where are you getting `pipefail` from there?

I think it’s [default with the `stdenv` builder.](https://github.com/NixOS/nixpkgs/blob/3e6461ff2de9badd8eb67a2e07b5747585544efd/pkgs/stdenv/generic/setup.sh#L5)

> Also, I see no such behavior from Xcode 15.1.

Me neither. I guess this is either consistent luck (the situation is inherently racy), or between the versions things got split in `xcodebuild` from one `printf` for both lines to one per line or something like that.

**Comment by jakubgs** [comment]
>I think it’s [default with the stdenv builder.](https://github.com/NixOS/nixpkgs/blob/3e6461ff2de9badd8eb67a2e07b5747585544efd/pkgs/stdenv/generic/setup.sh#L5)

Oh, interesting, that's good to know.

Well, I tested it with pipefail and can't reproduce it:
```
 > cat test.sh 
#!/usr/bin/env bash
set -ee
set -o pipefail
xcodebuild -version | head -n1

 > ./test.sh  
Xcode 15.1

 > echo $?
0
```
Can you with the newer version?

**Comment by mroi** [comment]
Thanks for the script. With the Xcode beta, I can even demonstrate it’s racy:
```
Michael@carpo:~/Desktop > cat test.sh 
#!/usr/bin/env bash
set -ee
set -o pipefail
xcodebuild -version | head -n1

Michael@carpo:~/Desktop > ./test.sh 
Xcode 16.0

Michael@carpo:~/Desktop > echo $?
0

Michael@carpo:~/Desktop > ./test.sh 
Xcode 16.0

Michael@carpo:~/Desktop > echo $?
141
```

**Comment by siddarthkay** [comment]
Indeed this seems to be a problem with Xcode 16 beta.
```
./test.sh
Xcode 16.0
echo $?
141
./test.sh
Xcode 16.0
echo $?
141
```

after switching locally to stable Xcode 15.4

```
./test.sh
Xcode 15.4
echo $?
0
./test.sh
Xcode 15.4
echo $?
0
./test.sh
Xcode 15.4
echo $?
0
```

**Comment by mroi** [comment]
Current Nixpkgs no longer contains the incriminated line.
This was fixed by 15d5ce3687725ed41921289a1ab8fdbe1a69e5e4.
